[CI Environment] Token Mechanism Uplift (support tokens as GitHub Secret + Migrate Settings.Json to Settings YAML) (#1516)
Azure/ResourceModules: docs/wiki/The CI environment - Token replacement.md

## (Optional) Local Custom Tokens

These are tokens defined in the Git Repository inside a [settings.yml](https://github.com/Azure/ResourceModules/blob/main/settings.yml) file. This allows creating tokens that are local and updatable via Source Control mechanisms. Here is an example on where these tokens are stored. You can add key-value pairs as required:

```yml
localToken_tokenA: 'foo'
localToken_tokenB: 'bar'

```

> > **Note:** The CI pipelines automatically removes `localToken_` section from the name when processing the tokens replacement. This means that your actual token name is `tokenA` and NOT `localToken_tokenA`.

Reviewer: ```suggestion
> **Note:** The CI pipelines automatically removes the `localToken_` prefix from the name when processing the tokens replacement. This means that your actual token name is `tokenA` and NOT `localToken_tokenA`.
```

Let's say you'd want to use this token inside a Key Vault parameter file, to deploy the Key Vault with a name that contains this token:

```json
"parameters": {
    "name": {
        "value": "<<tokenA>>-keyVault-<<tokenB>>"
    }
}
```

Once the Key Vault is deployed, you'll notice that the Key Vault name in Azure will be `foo-keyVault-bar`

> The token prefix `'<<'` and suffix `'>>'` in the above example are also configurable in the [settings.yml](https://github.com/Azure/ResourceModules/blob/main/settings.yml) file.

Reviewer: ```suggestion
The token prefix `'<<'` and suffix `'>>'` in the above example are also configurable in the [settings.yml](https://github.com/Azure/ResourceModules/blob/main/settings.yml) file and are used to identify the tokens in the files.
```

> The solution comes with one predefined local token `namePrefix`. This token is leveraged in most of the parameter files for test deployments. It allows using a consistent naming prefix that is applied to all resources being tested. There are two ways this token can be set and one will take precedence over the other:

Reviewer: ```suggestion
The solution comes with one predefined local token `namePrefix`. This token is leveraged in most of the parameter & test files for deployments. It allows using a consistent naming prefix that is applied to all resources being tested. There are two ways this token can be set and one will take precedence over the other:
```

1. By updating the value of `localToken_namePrefix` in the [settings.yml](https://github.com/Azure/ResourceModules/blob/main/settings.yml), which then becomes `namePrefix` when the pipelines run.

> 1. Creating a GitHub Secret / ADO variable called `TOKEN_NAMEPREFIX`, which then becomes `namePrefix` when the pipelines run. If created as a secret or ADO variable, it will only take precedence when the `localToken_namePrefix` in the [settings.yml](https://github.com/Azure/ResourceModules/blob/main/settings.yml) is left empty.

Reviewer: ```suggestion
1. Creating a GitHub Secret / ADO variable called `TOKEN_NAMEPREFIX`, which then becomes `namePrefix` when the pipelines run. If created as a secret or ADO variable, it is only applied if the `localToken_namePrefix` in the [settings.yml](https://github.com/Azure/ResourceModules/blob/main/settings.yml) is left empty.
```
